### PR TITLE
Refactored FIPS Checks into a Single Class (For Issue #34772)

### PR DESCRIPTION
**Pull request by LuisAlvelaMendes**
I have refactored all FIPS140 checks indicated in this issue: https://github.com/elastic/elasticsearch/issues/34772

Into a single FIPSChecks class, I also tried to implement a custom FIPSContext and interface that operate much like Bootstrap contexts, except without worrying about sometimes not being enforced (as FIPS is always enforced) and without requiring any metadata.

I also modified the unit tests to fit the changes and tried to call this class and the check method (that runs all checks) in the appropriate place, as discussed:

> A fitting place to call the checks is in createComponents of the Security plugin

### CODE REVIEW

**Review by jkakavas — CHANGES_REQUESTED** [review]
Hi @LuisAlvelaMendes !

Thanks once again for taking this up :) I did a first pass of your PR and I've left several comments and suggestions about how I think this could be further refactored. Let me know what you think. 

Please also make sure that you have configured your git installation correctly to set your name in the commits ( https://help.github.com/articles/setting-your-username-in-git/ )

Finally also make sure that you read through or [contribution guide](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md) and follow all the steps mentioned there. At the absolute minimum, ensure that [precommit check](https://github.com/elastic/elasticsearch/blob/master/TESTING.asciidoc#running-verification-tasks) is passing. 

Let me know if you have any questions

File: .gitignore [hunk 1]
```diff
@@ -41,3 +41,4 @@ html_docs
 # random old stuff that we should look at the necessity of...
 /tmp/
 eclipse-build
+/bin/
```
[the remark this answers is not in the document]
Comment (LuisAlvelaMendes) [reply .gitignore]: This is my mistake, sorry for including certain files that weren't relevant

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSInterface.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+import java.util.Objects;
+
+import org.elasticsearch.bootstrap.FIPSContext;
+import org.elasticsearch.env.Environment;
+
+/**
+ * Encapsulates a FIPS check.
+ */
+public interface FIPSInterface {
```
[the remark this answers is not in the document]
Comment (LuisAlvelaMendes) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSInterface.java]: The purpose would be to return an adequate FIPSCheckResult while also maintaining some consistency with how bootstrap checks were written .. I'm not quite sure what you mean b having the checks themselves implement it.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSChecks.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+				return check4;
+			}
+		}
+		
+		return null;
+	}
+	
+    public FIPSCheckResult keystoreCheck(FIPSContext context) {
+
+        if (XPackSettings.FIPS_MODE_ENABLED.get(context.settings)) {
```
[the remark this answers is not in the document]
Comment (LuisAlvelaMendes) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSChecks.java]: Do you mean add an if clause like `if (XPackSettings.FIPS_MODE_ENABLED.get(context.settings)) {}` surrounding it? I'm not sure what impact that would have

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSChecks.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+		FIPSCheckResult check1 = keystoreCheck(context);
+		FIPSCheckResult check2 = licenseCheck(context);
+		FIPSCheckResult check3 = passwordHashingAlgorithmCheck(context);
+		FIPSCheckResult check4 = secureSettingsCheck(context, env);
+		
+		if(check1.isSuccess() && check2.isSuccess() && check3.isSuccess() && check4.isSuccess()) {
+			return FIPSCheckResult.success();
+		}
+		
+		else {
```
[the remark this answers is not in the document]
Comment (LuisAlvelaMendes) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSChecks.java]: How do you suggest we do it then?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 5]
```diff
@@ -1063,10 +1066,12 @@ public void accept(DiscoveryNode node, ClusterState state) {
         @Override
         public void accept(DiscoveryNode node, ClusterState state) {
             if (inFipsMode) {
-                License license = LicenseService.getLicense(state.metaData());
-                if (license != null &&
-                    FIPS140LicenseBootstrapCheck.ALLOWED_LICENSE_OPERATION_MODES.contains(license.operationMode()) == false) {
-                    throw new IllegalStateException("FIPS mode cannot be used with a [" + license.operationMode() +
+
+                License.OperationMode licenseOperationMode = FIPSChecksInstance.get().getLicenseOperationMode();
```
[the remark this answers is not in the document]
Comment (LuisAlvelaMendes) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: How, exactly? I thought we had established the license operations should be independant from any metadata.

[the remark this answers is not in the document]
Comment (LuisAlvelaMendes) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: How, exactly? I thought we had established the license operations should be independant from any metadata.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 6]
```diff
@@ -499,6 +498,10 @@ protected Clock getClock() {
         securityActionFilter.set(new SecurityActionFilter(authcService.get(), authzService, getLicenseState(),
                 requestInterceptors, threadPool, securityContext.get(), destructiveOperations));
 
+        FIPSContext context = new FIPSContext(settings);
+        FIPSChecksInstance.set(new FIPSChecks(getLicenseState().getOperationMode()));
```
[the remark this answers is not in the document]
Comment (LuisAlvelaMendes) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: So we can get that instance with those values from the instruction `FIPSChecksInstance.get().check(context, env);` and call the checks

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSChecks.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSChecks.java]: I meant that we can have the if clause once in `Security.java` and call `check()` on all the checks (or the one `check()` if we decide to implement `FipsChecks` in a simpler manner ) IF `fips_mode` is set

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSInterface.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSInterface.java]: I was thinking of an interface named `FIPSCheck` that then the checks like `FipsLicencingCheck` or `FipsPasswordHashingAlgoCheck` would implement. But I do think we don't necessarily need this either, there is no mandate to simulate bootstrap checks implementation. Same with `FIPSCheckResult`, it feels like an unnecessary abstraction. WDYT ?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSChecks.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSChecks.java]: I was thinking something similar to how we use [addValidationError](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/ValidationException.java)

**Comment by jkakavas** [comment]
@LuisAlvelaMendes Thanks for the updates and apologies, I just noticed the new commits. I will take another look tomorrow, please feel free to ping me directly here every time there is a pending action on my side as otherwise your updates might get lost in the sea of Github notifications :)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java]: The question was about why is it necessary to wrap the `FIPSChecks` in a `SetOnce` and not do 
```java
FIPSContext context = new FIPSContext(settings);
FIPSChecks checks = new FIPSChecks(getLicenseState().getOperationMode())
checks.check(context, env);
```

**Comment by jkakavas** [comment]
Hi @LuisAlvelaMendes 

I thought I'd ping you and see how this is going and if you're still interested in seeing this PR through. If there's anything you'd want to discuss, feel free to add a comment here. 

https://github.com/elastic/elasticsearch/pull/36677/commits/e124f8a012165e2e60e217fd22f475e3c9b4d576 made me think that you had addressed some of the points we discussed above, hence my comment in https://github.com/elastic/elasticsearch/pull/36677#issuecomment-449084241,  but I then saw that this is the same as https://github.com/elastic/elasticsearch/pull/36677/commits/7a4e6ff574396a47b9c7d1d8a9870d6029fdd8f7 that you had committed earlier.

**Comment by LuisAlvelaMendes** [comment]
@jkakavas Sorry but due to some personal things I will not be able conclude this pull request.

**Comment by jkakavas** [comment]
Closing this in favor of #47499

**Pull request closed without merging**